### PR TITLE
JSDK-2356: quick start sample for room state change events 

### PR DESCRIPTION
**Pull request by makarandp0**
@manjeshbhargav, @aymenn 
This sample demonstrates the room state change event usage. It starts off by creating  a room with no media, and listens for the participant change event, as well as room state change events.   

Looks like:
room state

### CODE REVIEW

File: examples/roomevents/public/index.html [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      </div>
+      <div class="col-sm-6">
+        <div class="card-block">
+            <h5>  Room state is: </h5>
+            <div id="roomstate-unknown" class="roomstate unknown current">Unknown</div>
+            <div id="roomstate-connected" class="roomstate connected">Connected</div>
+            <div id="roomstate-disconnected" class="roomstate disconnected">Disconnected</div>
+            <div id="roomstate-reconnecting" class="roomstate reconnecting">Reconnecting</div>
+            <p>
+              Room once connected start off with `connected` state. Try turning off your network or WiFi connection, or switching to another Wifi network to see how Room state changes.
```
Comment (manjeshbhargav) [inline examples/roomevents/public/index.html]: @makarandp0 

> After you have joined the Room, please either turn off your internet network for a little while and then turn it back on, or switch between networks. The Twilio Video SDK will try to re-establish connection to the Room, which will transition to the "connecting" state. Once reconnection is complete, the Room will transition back to the "connected" state.

?

File: examples/roomevents/public/index.html [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            </p>
+        </div>
+        <input id="connectordisconnect" type="submit" class="btn btn-primary btn-block" value="Connect New User" style="display:none">
```
Comment (manjeshbhargav) [inline examples/roomevents/public/index.html]: @makarandp0 

> "Join Room" and "Leave Room"

?

File: examples/roomevents/src/helpers.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+var Video = require('twilio-video');
+
+/**
+ * Creates a Room and watches for room status events.
+ * @param {string} token - Token for joining the Room
+ * @param {function} onStateChange - Callback for state change.
+ * @returns {CancelablePromise<Room>}
+ */
+function createRoomAndUpdateOnStateChange(token, onStateChange) {
```
Comment (manjeshbhargav) [inline examples/roomevents/src/helpers.js]: @makarandp0 You do not need to demonstrate `connect()` here. You can just take in a Room and a state change callback as arguments:
```js
/**
 * Listen to Room reconnection events and update the UI accordingly.
 * @param {Room} room - The Room you have joined
 * @param {function} updateRoomState - Updates the app UI with the new Room state
 * @returns {void}
 */
function setupReconnectionUpdates(room, updateRoomState) {
  room.on('disconnected', (room, error) => {
    if (error.code === 20104) {
      console.log('Signaling reconnection failed due to expired AccessToken!');
    } else if (error.code === 53000) {
      console.log('Signaling reconnection attempts exhausted!');
    } else if (error.code === 53204) {
      console.log('Signaling reconnection took too long!');
    }
    updateRoomState(room.state);
  });
  room.on('reconnected', function() {
    console.log('Reconnected to the Room!');
    updateRoomState(room.state);
  });
  room.on('reconnecting', function(error) {
    if (error.code === 53001) {
      console.log('Reconnecting your signaling connection!', error.message);
    } else if (error.code === 53405) {
      console.log('Reconnecting your media connection!', error.message);
    }
    updateRoomState(room.state);
  });
}
```
?

File: package.json [hunk 4]
```diff
@@ -22,9 +23,12 @@
     "clean:examples:localvideofilter": "rimraf examples/localvideofilter/public/index.js examples/localvideofilter/public/helpers.js",
     "clean:examples:localvideosnapshot": "rimraf examples/localvideosnapshot/public/index.js examples/localvideosnapshot/public/helpers.js",
     "clean:examples:mediadevices": "rimraf examples/mediadevices/public/index.js examples/mediadevices/public/helpers.js",
+    "clean:examples:roomevents": "rimraf examples/roomevents/public/index.js examples/roomevents/public/helpers.js",
     "clean:examples:screenshare": "rimraf examples/screenshare/public/index.js examples/screenshare/public/helpers.js",
     "clean:quickstart": "rimraf quickstart/public/index.js",
-    "start": "npm run clean && npm run build && node server"
+    "start": "npm run clean && npm run build && node server",
+    "watch:roomevents": "copyfiles -f examples/roomevents/src/helpers.js examples/roomevents/public && watchify examples/roomevents/src/index.js -d -o examples/roomevents/public/index.js",
```
Comment (manjeshbhargav) [inline package.json]: @makarandp0 Let's not add watch events now. We can add it to all examples and quickstart in a separate PR.

File: server/index.js [hunk 5]
```diff
@@ -30,6 +30,7 @@ var app = express();
   'localvideofilter',
   'localvideosnapshot',
   'mediadevices',
+  'roomevents',
```
Comment (manjeshbhargav) [inline server/index.js]: @makarandp0 `reconnection`?

**Comment by makarandp0** [comment]
Thank you for taking a look and your comments @manjeshbhargav - I have incorporated your feedback  now. Please take another look.

File: examples/reconnection/public/index.html [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      </div>
+      <div class="col-sm-6">
+        <div class="card-block">
+            <h5>  Room state is: </h5>
+            <div class="roomstate unknown current">Unknown</div>
+            <div class="roomstate connected">Connected</div>
+            <div class="roomstate disconnected">Disconnected</div>
+            <div class="roomstate reconnecting">Reconnecting</div>
+            <p>
+              After you have joined the Room, please either turn off your internet network for a little while and then turn it back on, or switch between networks. The Twilio Video SDK will try to re-establish connection to the Room, which will transition to the "connecting" state. Once reconnection is complete, the Room will transition back to the "connected" state.
```
Comment (manjeshbhargav) [inline examples/reconnection/public/index.html]: @makarandp0 Nitpick: can you make this multiline?

File: examples/index.html [hunk 7]
```diff
@@ -121,6 +121,18 @@ <h4 class="card-header">
               </p>
             </div>
           </div>
+          <div class="card" style="text-decoration: none">
+            <a href="/roomevents">
+              <h4 class="card-header">
+                Room Events
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply examples/index.html]: @makarandp0 You forgot to change the title.

File: examples/reconnection/public/index.html [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+              Reconnection States and Events
+            </h4>
+            <pre class="language-javascript"></pre>
+          </div>
+        </div>
+      </div>
+      <div class="col-sm-6">
+        <div class="card-block">
+            <h5>  Room state is: </h5>
+            <div class="roomstate unknown current">Unknown</div>
```
Comment (manjeshbhargav) [inline examples/reconnection/public/index.html]: @makarandp0 You can get rid of this state.

File: examples/reconnection/src/helpers.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  room.on('disconnected', (room, error) => {
+    if (error.code === 20104) {
+      console.log('Signaling reconnection failed due to expired AccessToken!');
+    } else if (error.code === 53000) {
+      console.log('Signaling reconnection attempts exhausted!');
+    } else if (error.code === 53204) {
+      console.log('Signaling reconnection took too long!');
+    }
+    updateRoomState(room.state);
+  });
```
Comment (manjeshbhargav) [inline examples/reconnection/src/helpers.js]: @makarandp0 Nitpick: add extra line?

File: examples/reconnection/src/helpers.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  room.on('reconnected', function() {
+    console.log('Reconnected to the Room!');
+    updateRoomState(room.state);
+  });
```
Comment (manjeshbhargav) [inline examples/reconnection/src/helpers.js]: @makarandp0 Nitpick: add extra line?

File: examples/reconnection/src/helpers.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  room.on('reconnecting', function(error) {
+    if (error.code === 53001) {
+      console.log('Reconnecting your signaling connection!', error.message);
+    } else if (error.code === 53405) {
+      console.log('Reconnecting your media connection!', error.message);
+    }
+    updateRoomState(room.state);
+  });
+}
```
Comment (manjeshbhargav) [inline examples/reconnection/src/helpers.js]: @makarandp0 Nitpick: add extra line?

File: examples/reconnection/src/index.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+const getRoomCredentials = require('../../util/getroomcredentials');
+const getSnippet = require('../../util/getsnippet');
+const helpers = require('./helpers');
+const setupReconnectionUpdates = helpers.setupReconnectionUpdates;
+const connectOrDisconnect = document.querySelector('input#connectordisconnect');
+const mediaContainer = document.getElementById('remote-media');
+let roomName = null;
+let room = null;
+let someRoom = null;
+// let lastRoomState = "unknown";
```
Comment (manjeshbhargav) [inline examples/reconnection/src/index.js]: @makarandp0 Please clean up commented out code.

File: examples/reconnection/src/index.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  event.preventDefault();
+  return room ? disconnectFromRoom() : connectToRoom();
+}
+
+/**
+ * update the UI to indicate room state.
+ */
+function onRoomStateChange(newState) {
+  const oldStateBtn = document.querySelector('div.current');
+  // const oldStateBtn = document.getElementById("roomstate-" + lastRoomState);
```
Comment (manjeshbhargav) [inline examples/reconnection/src/index.js]: @makarandp0 Please clean up commented out code.

File: examples/reconnection/src/index.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  if (oldStateBtn) {
+    oldStateBtn.classList.remove('current');
+  }
+
+
+  // lastRoomState = someRoom.state;
```
Comment (manjeshbhargav) [inline examples/reconnection/src/index.js]: @makarandp0 Please clean up commented out code.

File: examples/reconnection/src/index.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * update the UI to indicate room state.
+ */
+function onRoomStateChange(newState) {
+  const oldStateBtn = document.querySelector('div.current');
+  // const oldStateBtn = document.getElementById("roomstate-" + lastRoomState);
+  if (oldStateBtn) {
+    oldStateBtn.classList.remove('current');
+  }
+
```
Comment (manjeshbhargav) [inline examples/reconnection/src/index.js]: @makarandp0 Remove unnecessary blank lines?

File: examples/reconnection/src/index.js [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+  const pre = document.querySelector('pre.language-javascript');
+
+  pre.innerHTML = Prism.highlight(snippet, Prism.languages.javascript);
+
+  // Get the credentials to connect to the Room.
+  const creds = await getRoomCredentials();
+
+  // Connect to a random Room with no media. This Participant will
+  // display the media of the other Participants that will enter
+  // the Room and watch for dominant speaker updates.
```
Comment (manjeshbhargav) [inline examples/reconnection/src/index.js]: @makarandp0 `... and watch for reconnection updates.`?

File: examples/reconnection/src/index.js [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+  someRoom = await Video.connect(creds.token, {});
```
Comment (manjeshbhargav) [inline examples/reconnection/src/index.js]: @makarandp0 If you want to join without media, you have to explicitly provide an empty `tracks` array:
```js
connect(creds.token, { tracks: [] });
```
You may want to update your dominant speaker example with this change.

**Comment by manjeshbhargav** [comment]
@makarandp0 One more thing: I noticed that, if you turn off the wi-fi until the Room reaches the "disconnected" state, it looks like we can no longer repeat the example because the button still says "Leave Room" and the <video> element doesn't go away. I think this is because `someRoom` is only created once, and once it is disconnected, it stays disconnected forever. Can you make sure the example can be used even after `someRoom` gets disconnected?

File: examples/reconnection/src/index.js (hunk 17 above)
Comment (makarandp0) [reply examples/reconnection/src/index.js]: Thanks for noticing this.

**Comment by makarandp0** [comment]
Thanks @manjeshbhargav, Good point about being able to reuse the sample after room disconnected. I have not made some changes so that user can recreate the room after disconnected state.

File: examples/dominantspeaker/src/helpers.js [hunk 18]
```diff
@@ -6,13 +6,12 @@ var Video = require('twilio-video');
...
  * @param {string} token - Token for joining the Room
- * @param {string} roomName - Room name
  * @returns {CancelablePromise<Room>}
  */
-function connectToRoomWithDominantSpeaker(token, roomName) {
+function connectToRoomWithDominantSpeaker(token) {
   return Video.connect(token, {
     dominantSpeaker: true,
-    name: roomName
+    tracks: []
```
Comment (manjeshbhargav) [inline examples/dominantspeaker/src/helpers.js]: @makarandp0 I think the code snippet that is displayed in the app should not have the empty tracks. We should pass empty tracks to the Participant that only listens to the events.

Comment (makarandp0) [reply examples/dominantspeaker/src/helpers.js]: @manjeshbhargav, This call is made by participant that only watches the tracks (the one that creates `someroom`. But I agree, that its misleading to have empty tracks in the example. It does not hurt if we created someroom is created with the tracks. I will remove the tracks option from here, and merge it in.

**Pull request merged**